### PR TITLE
Some more minor flush optimizations

### PR DESCRIPTION
**Pull request by louiscryan**
Reduce number of explicit flushes after releasing bytes to flow control
Reduce use of explicit flush for cases where a subsequent flush is known to follow immediately
Add test to validate the number of physical writes per call type

### CODE REVIEW

**Comment by louiscryan** [comment]
@ejona86 @nmittler

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 1]
```diff
@@ -293,7 +294,14 @@ private void scheduleNextChunk() {
         Chunk nextChunk = chunks.peek();
         if (nextChunk != null) {
           scheduled = true;
-          executor.schedule(dispatchTask, nextChunk.delayMicroseconds, TimeUnit.MICROSECONDS);
+          if (executor instanceof ScheduledExecutorService) {
```
Comment (ejona86) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: && nextChunk.delayMicroseconds == 0

??

File: netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java [hunk 2]
```diff
@@ -238,20 +238,16 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
     } else if (msg instanceof RequestMessagesCommand) {
       ((RequestMessagesCommand) msg).requestMessages();
     } else {
-      AssertionError e =
-          new AssertionError("Write called for unexpected type: " + msg.getClass().getName());
-      ReferenceCountUtil.release(msg);
-      promise.setFailure(e);
-      throw e;
+      throw new AssertionError("Write called for unexpected type: " + msg.getClass().getName());
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java]: Why this change?

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 1 above)
Comment (louiscryan) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: not sure it matters, scheduling with a delay of 0 should be the same as execute

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java [hunk 3]
```diff
@@ -31,10 +31,17 @@
...
+import io.grpc.MethodType;
 import io.grpc.Status;
 
 /** An observer of server-side stream events. */
 public interface ServerStreamListener extends StreamListener {
+
+  /**
+   * The {@link MethodType} for the call.
+   */
+  MethodType getMethodType();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: This is strange. I'd prefer to not have it, and it seems unnecessary.

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java [hunk 4]
```diff
@@ -89,7 +89,7 @@ protected void inboundDeliveryPaused() {
   protected void internalSendHeaders(Metadata.Headers headers) {
     writeQueue.enqueue(new SendResponseHeadersCommand(id(),
         Utils.convertServerHeaders(headers), false),
-        true);
+        !getMethodType().serverSendsOneMessage());
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Instead of this, could sendHeaders() have a flush argument, and this logic is done in ServerImpl.ServerCallImpl?

File: stub/src/main/java/io/grpc/stub/ServerCalls.java [hunk 5]
```diff
@@ -76,9 +76,6 @@ public void onPayload(ReqT request) {
               // We delay calling method.invoke() until onHalfClose(), because application may call
               // close(OK) inside invoke(), while close(OK) is not allowed before onHalfClose().
               this.request = request;
-
-              // Request delivery of the next inbound message.
-              call.request(1);
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: Hmm... So before if the type was wrong we would end up failing the call. Now we will deadlock. I don't know how I feel about that yet...

File: stub/src/main/java/io/grpc/stub/Calls.java [hunk 6]
```diff
@@ -172,10 +172,11 @@
       ReqT param,
       Call.Listener<RespT> responseListener) {
     call.start(responseListener, new Metadata.Headers());
-    call.request(1);
     try {
       call.sendPayload(param);
       call.halfClose();
+      // Request after sending the payload otherwise we prematurely trigger a flush.
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/Calls.java]: Is this still true after your change to conditionally call flush after returnProcessedBytes?

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 1 above)
Comment (ejona86) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Hmm... my condition was a bit off. I'm concerned about the delay not happening.

How about something more along these lines:
if (nextChunk.delayMicroseconds != 0) {
  ((ScheduledExecutorService) executor).schedule(
      dispatchTask, nextChunk.delayMicroseconds, TimeUnit.MICROSECONDS);
} else {
  dispatchTask.run();
}

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java (hunk 3 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: Yeah, I recall making a similar comment on a previous PR.

File: netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java (hunk 2 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java]: +1 ... don't we need to at least release the message?

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 1 above)
Comment (louiscryan) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Now that we're allowing construction with an executor that is not a ScheduledExecutorService we can't assume the cast here. See the transport efficiency test use of direct executor for an example

File: netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java (hunk 2 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java]: sorry, broken 3-way merge. Fixed

File: netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java (hunk 4 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java]: Yup, that works. Done

File: stub/src/main/java/io/grpc/stub/Calls.java (hunk 6 above)
Comment (louiscryan) [inline stub/src/main/java/io/grpc/stub/Calls.java]: A flush triggered by the call to 'request' can race with the flush in onReadComplete if the server responds quickly.

I have a change in mind that will make this moot anyway... Construct the deframer with pending messages == 1 rather than calling 'request' at the start of each call, this avoids a context switch for non-direct & blocking clients

File: core/src/main/java/io/grpc/transport/ServerStreamListener.java (hunk 3 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/ServerStreamListener.java]: It is ugly, removed with Eric's suggested change to ServerCallImpl

File: stub/src/main/java/io/grpc/stub/ServerCalls.java (hunk 5 above)
Comment (louiscryan) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: Yeah, it's somewhat related to https://github.com/grpc/grpc-java/issues/433

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 1 above)
Comment (ejona86) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: With my suggestion we still construct with a ScheduledExecutorService, but we always get the benefit of direct when delaying is unnecessary. That seems like a win overall.

Comment (louiscryan) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Done

**Comment by louiscryan** [comment]
PTAL

File: stub/src/main/java/io/grpc/stub/ServerCalls.java (hunk 5 above)
Comment (louiscryan) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: I could go either way on this one. Assuming we do something to pipeline window updates as discussed in the other bug I'd be happy to revert this.

Comment (louiscryan) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: I've reverted this for the moment. We can revisit later

**Comment by ejona86** [comment]
LGTM, but I haven't reviewed NettyTransportIoEfficiencyTest.

:( You rebased. That makes it hard to see what changed since previous reviews. I guess it's good that the majority of the non-obvious changes were additions.

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        .setResponseSize(payloadSize)
+        .setResponseType(Messages.PayloadType.COMPRESSABLE)
+        .setPayload(Messages.Payload.newBuilder()
+            .setBody(ByteString.copyFrom(new byte[payloadSize])))
+        .build();
+    blockingStub.unaryCall(request);
+    int maxWindowUpdates = (payloadSize / (65535 / 2));
+    // We use bounds checking as we don't have precisely predictable behavior here. The number
+    // of DATA frames read in a unit depends on how big a chunk we get from the IO layer.
+    System.err.println("ClientCountingChannel.flushCount " + ClientCountingChannel.flushCount);
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Get rid of these printlns?

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+    System.err.println("ServerCountingChannel.flushCount " + ServerCountingChannel.flushCount);
+    assertTrue(maxWindowUpdates + payloadWrites >= ClientCountingChannel.flushCount);
+    // +1 here for headers
+    assertTrue(maxWindowUpdates + payloadWrites + 1 >= ServerCountingChannel.flushCount);
+  }
+
+  @Test(timeout = 10000)
+  public void largeUnaryWithWindowUpdateAndDirectExecutor() throws Exception {
+    System.err.println("START");
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Git rid of this println, and the one in the finally?

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+          .negotiationType(NegotiationType.PLAINTEXT);
+      if (directExecutor) {
+        builder.executor(MoreExecutors.newDirectExecutorService());
+      }
+      channel = builder.build();
+      blockingStub = TestServiceGrpc.newBlockingStub(channel);
+      asyncStub  = TestServiceGrpc.newStub(channel);
+      blockingStub.unaryCall(SMALL_REQUEST);
+      // Need to sleep so that SETTINGS_ACK is received and processed as it can cause a flush.
+      Thread.sleep(1000);
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Ouch. That means each of these tests will take at least a second. That greatly increases our testing time.

Can we instead wait for the flush count to increase due to the SETTINGS ACK? Hmm... Unfortunate all around...

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+      flushCount++;
+      super.doWrite(in);
+    }
+  }
+
+
+
+  public static class ServerSocketChannel extends NioServerSocketChannel {
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: What does this provide over just `NioServerSocketChannel`?

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+
+  public static class ClientCountingChannel extends NioSocketChannel {
+
+    public static volatile int flushCount = 0;
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+      flushCount++;
+      //Thread.dumpStack();
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: Remove?

File: core/src/main/java/io/grpc/ChannelImpl.java [hunk 12]
```diff
@@ -332,7 +329,7 @@ public void sendPayload(ReqT payload) {
       // For unary requests, we don't flush since we know that halfClose should be coming soon. This
       // allows us to piggy-back the END_STREAM=true on the last payload frame without opening the
       // possibility of broken applications forgetting to call halfClose without noticing.
-      if (!unaryRequest) {
+      if (!method.getType().clientSendsOneMessage()) {
```
Comment (sonyakc) [inline core/src/main/java/io/grpc/ChannelImpl.java]: rather than check that client doesn't send one message, I'd suggest the following to avoid the negated ```if``` check

``` if (method.getType().clientSendsMultipleMessages()) { ```

File: integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 13]
```diff
@@ -293,7 +293,14 @@ private void scheduleNextChunk() {
         Chunk nextChunk = chunks.peek();
         if (nextChunk != null) {
           scheduled = true;
```
Comment (sonyakc) [inline integration-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: ```scheduled = true;``` should be moved inside the ```else {...}``` since that's where the job is actually getting scheduled using the executor. Immediate execution shouldn't result in scheduled being set to true, correct? As per the previous code, it was always scheduled hence setting true prior to scheduling the task.

File: integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java [hunk 14]
```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright 2015, Google Inc. All rights reserved.
```
Comment (sonyakc) [inline integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java]: 2018 copyright year?

**Comment by ejona86** [comment]
Closing as this is quite old and would have to be redone. https://github.com/grpc/grpc-java/issues/4884 is tracking this, with https://github.com/grpc/grpc-java/pull/9273 having done part of what was in here.
